### ISSUE TITLE
TypeError: initializer for ctype 'char *' must be a str or list or tuple, not unicode

### ISSUE THREAD

**Issue by jkroepke** [letsencrypt/letsencrypt#2360]
Hi,

I try to use the /root/.local/share/letsencrypt/bin/letsencrypt-renewer, but this error happend.
I'm on ubuntu 14.04.

Currently, master is checkout. The error happend on tag v0.3.0 and v0.2.0, too

```
Processing domain.conf
2016-02-04 00:53:53,640:INFO:requests.packages.urllib3.connectionpool:Starting new HTTPS connection (1): acme-v01.api.letsencrypt.org
/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py:791: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
  InsecureRequestWarning)
2016-02-04 00:53:54,298:INFO:letsencrypt.crypto_util:Generating key (2048 bits): /etc/letsencrypt/keys/0054_key-letsencrypt.pem
Traceback (most recent call last):
  File "/root/.local/share/letsencrypt/bin/letsencrypt-renewer", line 11, in <module>
    sys.exit(main())
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt/renewer.py", line 203, in main
    renew(cert, old_version)
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt/renewer.py", line 97, in renew
    new_certr, new_chain, new_key, _ = le_client.obtain_certificate(sans)
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt/client.py", line 264, in obtain_certificate
    csr = crypto_util.init_save_csr(key, domains, self.config.csr_dir)
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt/crypto_util.py", line 78, in init_save_csr
    csr_pem, csr_der = make_csr(privkey.pem, names)
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/letsencrypt/crypto_util.py", line 118, in make_csr
    value=", ".join("DNS:%s" % d for d in domains)
  File "/root/.local/share/letsencrypt/local/lib/python2.7/site-packages/OpenSSL/crypto.py", line 651, in __init__
    extension = _lib.X509V3_EXT_nconf(_ffi.NULL, ctx, type_name, value)
TypeError: initializer for ctype 'char *' must be a str or list or tuple, not unicode

```

```
/root/.local/share/letsencrypt/bin/pip list
acme (0.3.0)
cffi (1.5.0)
ConfigArgParse (0.10.0)
configobj (5.0.6)
cryptography (1.2.2)
enum34 (1.1.2)
funcsigs (0.4)
idna (2.0)
ipaddress (1.0.16)
letsencrypt (0.3.0)
letsencrypt-apache (0.3.0)
mock (1.3.0)
ndg-httpsclient (0.4.0)
parsedatetime (1.5)
pbr (1.8.1)
pip (8.0.2)
psutil (3.4.2)
pyasn1 (0.1.9)
pycparser (2.14)
pyOpenSSL (0.15.1)
pyRFC3339 (1.0)
python-augeas (0.5.0)
python2-pythondialog (3.3.0)
pytz (2015.7)
requests (2.9.1)
setuptools (19.7)
six (1.10.0)
Werkzeug (0.11.3)
zope.component (4.2.2)
zope.event (4.1.0)
zope.interface (4.1.3)

```

**Comment by jkroepke** [comment]
Maybe: https://github.com/pyca/pyopenssl/issues/15

**Comment by Athanasius** [comment]
I'm also seeing this issue.  I'll check into what jkroepke pointed out to see if there's some small tweak that can be made.

**Comment by Athanasius** [comment]
This can be fixed with a one-line tweak to renewer.py, as below.  Obviously the correct thing to do here would likely be to punycode the domain name instead to handle those using non-ASCII ?

--- /root/tmp/renewer.py        2016-02-09 16:59:46.749480907 +0000
+++ /root/.local/share/letsencrypt/lib/python2.7/site-packages/letsencrypt/renewer.py   2016-02-09 16:59:48.849461586 +0000
@@ -94,6 +94,7 @@
     le_client = client.Client(config, acc, authenticator, None)
     with open(cert.version("cert", old_version)) as f:
         sans = crypto_util.get_sans_from_cert(f.read())
+    sans = [s.encode('ascii', 'ignore') for s in sans]
     new_certr, new_chain, new_key, _ = le_client.obtain_certificate(sans)
     if new_chain:
         # XXX: Assumes that there was a key change.  We need logic

**Comment by Athanasius** [comment]
Oh, and why /root/tmp/renewer.py for the 'before' side of that diff?  Because I couldn't find where the hell the equivalent file is in the git clone.  Makes me wonder just how screwy the installation mechanism is....

**Comment by schoen** [comment]
> Oh, and why /root/tmp/renewer.py for the 'before' side of that diff?  Because I couldn't find where the hell the equivalent file is in the git clone.  Makes me wonder just how screwy the installation mechanism is....

I'm sorry for the inconvenience; the reason that you can't find renewer.py
in the git master branch is that it's been deprecated and removed for
the upcoming release, replaced by `letsencrypt renew` (which incidentally
has a separate fix for the Unicode problem).

**Comment by Athanasius** [comment]
Ah, well that explains that then.  In which case ./letsencrypt-auto needs fixing, as I just did a pull from the git master and re-ran that command before starting work on renewals today.  Yet I still have letsencrypt-renewer and no 'renew' listed on ./letsencrypt --help

**Comment by jkroepke** [comment]
When `letsencrypt renew` is available on master branch?

**Comment by schoen** [comment]
> When `letsencrypt renew` is available on master branch?

It's available on the master branch now.

**Comment by jkroepke** [comment]
Thanks your the information. But the master branch will install letsencrypt 0.3.0 on the virtual env, so it is not availble for me...

https://github.com/letsencrypt/letsencrypt/blob/master/letsencrypt-auto-source/letsencrypt-auto#L647

**Comment by Athanasius** [comment]
As I said, letsencrypt-auto needs fixing.

On that note, I had already done a fresh 'git pull origin' and another run of ./letsencrypt-auto, but when I nuked ~/.local/share/letsencrypt and did that all again I had extra Debian packages installed, and presumably other more up to date things installed.  It basically looks like re-running letsencrypt-auto alone is NOT enough to get a fresh install.

Not the most user-friendly process at all (although I appreciate the effort to get it working even on Debian Wheezy, yes, I know I need to schedule the upgrade to current stable).

**Comment by schoen** [comment]
@jkroepke, current master now refers to 0.4.0 (the release has just happened).

I believe the instructions at https://github.com/letsencrypt/letsencrypt/blob/master/docs/contributing.rst describe how to run a more recent version of the development tree than the latest release.

@Athanasius, there's a substantially improved `letsencrypt-auto` in the latest release; I hope you're happier with it.

Because the 0.4.0 release is out now and replaces `letsencrypt-renewer` with `letsencrypt renew` (and should fix the Unicode issue), I'm going to close this. Thanks to @Athanasius for your interim suggestion and my apologies once again for the inconvenience.
